test/topology_custom/test_ip_mappings.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+import asyncio
+from test.pylib.manager_client import ManagerClient
+
+import pytest
+import logging
+
+from test.pylib.rest_client import inject_error_one_shot
+
+logger = logging.getLogger(__name__)
+
+@pytest.mark.asyncio
+async def test_broken_bootstrap(manager: ManagerClient):
+    server_a = await manager.server_add()
+    server_b = await manager.server_add(start=False)
+
+    await manager.cql.run_async("CREATE KEYSPACE test WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1}")
+    await manager.cql.run_async("CREATE TABLE test.test (a int PRIMARY KEY, b int)")
+    for i in range(100):
+        await manager.cql.run_async(f"INSERT INTO test.test (a, b) VALUES ({i}, {i})")
+    await inject_error_one_shot(manager.api, server_a.ip_addr, "crash-before-bootstrapping-node-added")
+    try:
+        await manager.server_start(server_b.server_id, timeout=5)
+        pytest.fail("Expected server_add to fail")
+    except Exception:
+        pass
+
+    await manager.server_stop(server_b.server_id)
+    await manager.server_stop(server_a.server_id)
+
+    stop_event = asyncio.Event()
+    async def worker():
+        logger.info("Worker started")
+        while not stop_event.is_set():
+            for i in range(100):
+                await manager.cql.run_async(f"INSERT INTO test.test (a, b) VALUES ({i}, {i})")
+                response = await manager.cql.run_async(f"SELECT * FROM test.test WHERE a = {i}")
+                assert response[0].b == i
+            await asyncio.sleep(0.1)
+        logger.info("Worker stopped")
+
+    await manager.server_start(server_a.server_id)
+    await manager.driver_connect()
+
+    worker_task = asyncio.create_task(worker())
```

Comment:
If I'm not wrong, we should always wait for CQL before querying the restarted servers. Tests that don't it can be flaky and fail, for example, 1/1000 times. I wouldn't risk it here.

Maybe I'm wrong, and we don't need it here. @kbr-scylla knows this stuff well.

Reply:
Agreed, better to be safe now than sorry later. We know from experience that `wait_for_cql_and_get_hosts` prevents certain type of flakiness

Reply:
Guys do we really have nothing else to do instead of nit picking on a simple test? There are no comments here that warrant me wasting half a day on a context switch to this old forgotten branch. The test run at least 500 times without `wait_for_cql_and_get_hosts` so without a definite evidence that it is needed here why would I waste time adding it? Rest of the comment are of the same importance.  If the test fails in 20 seconds it is not the end for the world since it should 100% pass anyway. Hanging for an hour in case of a failure is a problem worth fixing, not this.

Reply:
> If the test fails in 20 seconds it is not the end for the world since it should 100% pass anyway.

It will also sleep for 20 seconds unnecessarily in successful runs. Slows down CI runs for everyone for no reason. Multiply by all CI runs we're doing and that's a lot of time wasted for everyone.

Reply:
> There are no comments here that warrant me wasting half a day on a context switch to this old forgotten branch. 

It shouldn't take more than 20 minutes to address these comments. And you can do it in the morning the next day before daily to avoid disruptive context switching.

> The test run at least 500 times without wait_for_cql_and_get_hosts so without a definite evidence that it is needed here why would I waste time adding it?

Fine. Maybe it's not needed anymore with never python-driver versions. We'll find out.

Reply:
> > If the test fails in 20 seconds it is not the end for the world since it should 100% pass anyway.
> 
> It will also sleep for 20 seconds unnecessarily in successful runs. Slows down CI runs for everyone for no reason. Multiply by all CI runs we're doing and that's a lot of time wasted for everyone.

The test runs traffic for 20 seconds. What else do you propose to do? If you have something else in mind send the patch.

Reply:
Run traffic for 5 seconds? Or just do a pre-determined number of writes? I would expect that the first couple of writes would already reproduce the issue since every write needs the mapping.  Is running 20 seconds of workload really needed to reproduce the issue reliably?

Reply:
Any of that is not less arbitrary but now the review comes down to bike shedding about 20 vs 5, while the patch could have been in the next for a couple hours already.

Reply:
Our goal is not to push as many patches as we can in shortest amount of time.

There is clear value in keeping tests short as possible. Best would be if we could make this test not time dependent (sleep for X) but just do e.g. 100 writes and if there's a 99% chance that one of these writes will hit the no mapping issue then it's sufficient, and the writing part will take few milliseconds in most runs.

Reply:
> Our goal is not to push as many patches as we can in shortest amount of time.

There should be a balance. Bike shedding 5s vs 20s is just not it given that we do want some timeout anyway. It saturates developers time with unimportant stuff. I can nitpick any non trivial patch for eternity. There is no pride in it.

Reply:
Enough with the 5s vs 20s.

About this:
> Best would be if we could make this test not time dependent (sleep for X) but just do e.g. 100 writes and if there's a 99% chance that one of these writes will hit the no mapping issue then it's sufficient, and the writing part will take few milliseconds in most runs.

You still did not explain why you decided to go with the solution of "run workload for X seconds in background and then check" instead of "do N inserts in foreground" -- which is both faster and results in simpler code.

I checked the second version, works reliably (100/100 runs) with N = 10000. Compare:
```python
    stop_event = asyncio.Event()
    async def worker():
        logger.info("Worker started")
        while not stop_event.is_set():
            for i in range(100):
                await manager.cql.run_async(f"INSERT INTO test.test (a, b) VALUES ({i}, {i})")
                response = await manager.cql.run_async(f"SELECT * FROM test.test WHERE a = {i}")
                assert response[0].b == i
            await asyncio.sleep(0.1)
        logger.info("Worker stopped")

    await manager.server_start(server_a.server_id)
    await manager.driver_connect()

    worker_task = asyncio.create_task(worker())

    await asyncio.sleep(20)
    stop_event.set()
    await worker_task
```

with:
```python
    await manager.server_start(server_a.server_id)
    await manager.driver_connect()

    for i in range(10000):
        await manager.cql.run_async(f"INSERT INTO test.test (a, b) VALUES ({i}, {i})")
        response = await manager.cql.run_async(f"SELECT * FROM test.test WHERE a = {i}")
        assert response[0].b == i
```

The first one is unnecessarily complex and makes the test run 20 seconds longer.

It's not a nitpick, it would be a significant simplification to the code and improvement for our test suite performance.

(And do you want me to point you to some of reviews you did on my code because you had a simpler design in mind? And I didn't complain and call your suggestions "nitpicks")

Reply:
> You still did not explain why you decided to go with the solution of "run workload for X seconds in background and then check" instead of "do N inserts in foreground" -- which is both faster and results in simpler code.

Because I did not wrote the test and I do not strive for perfection in a test for a very specific problem. Would I wrote it differently? Probably. But why would I write a new test if one is a;ready exist.

and for me if the existing test do the job is good enough. I do not strive for code perfection in a test for a very specific problem.

> 
> (And do you want me to point you to some of reviews you did on my code because you had a simpler design in mind? And I didn't complain and call your suggestions "nitpicks")

About a test? Please show me.

Reply:
We agreed on daily to address this in a follow-up.